db/db_impl/db_impl_open.cc

```diff
@@ -1298,7 +1298,7 @@ Status DBImpl::RecoverLogFiles(const std::vector<uint64_t>& wal_numbers,
           flushed = true;
 
           cfd->CreateNewMemtable(*cfd->GetLatestMutableCFOptions(),
-                                 *next_sequence);
+                                 *next_sequence - 1);
```

Comment:
https://github.com/facebook/rocksdb/blob/f42e70bf561d4be9b6bbe7316d1c2c0c8a3818e6/db/memtable.h#L410-L411

Won't `*next_sequence - 1` be in an older (earlier) memtable?

Reply:
Yes, the last one used I think, this was how it was in `db/db_impl.cc` when the test last worked in v4.8

https://github.com/facebook/rocksdb/compare/4.8.fb...4.9.fb#diff-6fdb755f590d9b01ecb89bd8ceb28577e85536d4472f8e4fc3addeb9a65f3645L1393
 
Note the `- 1`.

      const SequenceNumber last_seq = WriteBatchInternal::Sequence(&batch) +
                                      WriteBatchInternal::Count(&batch) - 1;
      if ((*max_sequence == kMaxSequenceNumber) || (last_seq > *max_sequence)) {
        *max_sequence = last_seq;
      }
 
      ...

          cfd->CreateNewMemtable(*cfd->GetLatestMutableCFOptions(),
                                 *max_sequence);
 
The changes for 2pc between v4.8 and v4.9 are extensive, and I have not spent the time to understand enough of them to know if the `- 1` was "correct" in v4.8, and if the lack of a `- 1` after v4.8 is not "correct".

The change to subtract 1 does fix the `OptimisticTransactionTest.SequenceNumberAfterRecoverLargeTest` I added to this PR, note the test fails under TSAN (see below), and all other tests continue to pass.

This comment from https://github.com/facebook/rocksdb/pull/1458#issuecomment-259507341 (which was for the fix for `OptimisticTransactionTest.SequenceNumberAfterRecoverTest`) may help:

> "It's an off-by-one error. After recovery when the first memtable is created a sequence id 1 too large was used to create the memtable table. This sequence id is used for pessimistic transactions to determine if there is a conflict since start of transaction. Because of the OBO error the sequence id of the first insert into this memtable was smaller than the sequence id of the table so the transaction layer incorrectly thought it could not do conflict checking."

TSAN failure from https://app.circleci.com/pipelines/github/facebook/rocksdb/32352/workflows/b7891bee-651a-49cf-8ea4-de6103fcc444/jobs/666968

```
ThreadSanitizer: CHECK failed: sanitizer_deadlock_detector.h:67 "((n_all_locks_)) < (((sizeof(all_locks_with_contexts_)/sizeof((all_locks_with_contexts_)[0]))))" (0x40, 0x40) (tid=2071392)
    #0 __tsan::CheckUnwind() <null> (optimistic_transaction_test+0x4abddd)
    #1 __sanitizer::CheckFailed(char const*, int, char const*, unsigned long long, unsigned long long) <null> (optimistic_transaction_test+0x437864)
    #2 __sanitizer::DeadlockDetectorTLS<__sanitizer::TwoLevelBitVector<1ul, __sanitizer::BasicBitVector<unsigned long> > >::addLock(unsigned long, unsigned long, unsigned int) <null> (optimistic_transaction_test+0x42b386)
    #3 __sanitizer::DD::MutexAfterLock(__sanitizer::DDCallback*, __sanitizer::DDMutex*, bool, bool) <null> (optimistic_transaction_test+0x42a87d)
    #4 __tsan::MutexPostLock(__tsan::ThreadState*, unsigned long, unsigned long, unsigned int, int) <null> (optimistic_transaction_test+0x4b9010)
    #5 pthread_mutex_lock <null> (optimistic_transaction_test+0x46d72f)
    #6 rocksdb::port::Mutex::Lock() /root/project/port/port_posix.cc:80:23 (librocksdb.so.8.7+0x805e4e)
    #7 rocksdb::MutexLock::MutexLock(rocksdb::port::Mutex*) /root/project/./util/mutexlock.h:37:62 (librocksdb.so.8.7+0x962384)
    #8 rocksdb::ThreadLocalPtr::StaticMeta::GetThreadLocal() /root/project/util/thread_local.cc:359:17 (librocksdb.so.8.7+0x962384)
    #9 rocksdb::ThreadLocalPtr::StaticMeta::Swap(unsigned int, void*) /root/project/util/thread_local.cc:395:15 (librocksdb.so.8.7+0x962681)
    #10 rocksdb::ThreadLocalPtr::Swap(void*) /root/project/util/thread_local.cc:507:60 (librocksdb.so.8.7+0x96384c)
    #11 rocksdb::ColumnFamilyData::GetThreadLocalSuperVersion(rocksdb::DBImpl*) /root/project/db/column_family.cc:1237:26 (librocksdb.so.8.7+0x3a6e27)
    #12 rocksdb::DBImpl::GetAndRefSuperVersion(rocksdb::ColumnFamilyData*) /root/project/db/db_impl/db_impl.cc:4294:15 (librocksdb.so.8.7+0x46745a)
    #13 rocksdb::DBImpl::GetAndRefSuperVersion(unsigned int) /root/project/db/db_impl/db_impl.cc:4306:10 (librocksdb.so.8.7+0x46745a)
    #14 rocksdb::TransactionUtil::CheckKeysForConflicts(rocksdb::DBImpl*, rocksdb::LockTracker const&, bool) /root/project/utilities/transactions/transaction_util.cc:163:33 (librocksdb.so.8.7+0xaa5957)
    #15 rocksdb::OptimisticTransaction::CommitWithParallelValidate() /root/project/utilities/transactions/optimistic_transaction.cc:138:14 (librocksdb.so.8.7+0xa7ebd2)
    #16 rocksdb::OptimisticTransaction::Commit() /root/project/utilities/transactions/optimistic_transaction.cc:68:14 (librocksdb.so.8.7+0xa7e81d)
    #17 rocksdb::OptimisticTransactionTest_SequenceNumberAfterRecoverLargeTest_Test::TestBody() /root/project/utilities/transactions/optimistic_transaction_test.cc:1656:20 (optimistic_transaction_test+0x4ef419)
```
 
Potentially a false positive because of the recursive lock limit of 64 in https://github.com/llvm-mirror/compiler-rt/blob/master/lib/sanitizer_common/sanitizer_deadlock_detector.h#L117 , see https://github.com/google/sanitizers/issues/950 .

Reply:
Thanks. The API doc is already wrong for live writes (`SwitchMemtable()`). Your change would match that behavior.

The incorrect API doc is repeated here in the public API - https://github.com/facebook/rocksdb/blob/8acf17002a75727e036ad5974ddd335ba092a9be/include/rocksdb/listener.h#L484-L485

Despite that, the transaction code understands what `earliest_seq` really means - https://github.com/facebook/rocksdb/blob/8acf17002a75727e036ad5974ddd335ba092a9be/utilities/transactions/transaction_util.cc#L84-L85

I think for now this change is fine. We should reconsider the name and contract for `earliest_seq` though.